authority/tls.go

```diff
@@ -458,6 +463,91 @@ func (a *Authority) revokeSSH(crt *ssh.Certificate, rci *db.RevokedCertificateIn
 	return a.db.Revoke(rci)
 }
 
+// GenerateCertificateRevocationList returns a DER representation of a signed CRL.
+// It will look for a valid generated CRL in the database, check if it has expired, and generate
+// a new CRL on demand if it has expired (or a CRL does not already exist).
+//
+// force set to true will force regeneration of the CRL regardless of whether it has actually expired
+func (a *Authority) GenerateCertificateRevocationList(force bool) ([]byte, error) {
+
+	// check for an existing CRL in the database, and return that if its valid
+	crlInfo, err := a.db.GetCRL()
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !force && crlInfo != nil && crlInfo.ExpiresAt.After(time.Now().UTC()) {
+		return crlInfo.DER, nil
+	}
+
+	// some CAS may not implement the CRLGenerator interface, so check before we proceed
+	caCRLGenerator, ok := a.x509CAService.(casapi.CertificateAuthorityCRLGenerator)
+
+	if !ok {
+		return nil, errors.Errorf("CRL Generator not implemented")
+	}
+
+	revokedList, err := a.db.GetRevokedCertificates()
+
+	// Number is a monotonically increasing integer (essentially the CRL version number) that we need to
+	// keep track of and increase every time we generate a new CRL
+	var n int64 = 0
+	var bn big.Int
+
+	if crlInfo != nil {
+		n = crlInfo.Number + 1
+	}
+	bn.SetInt64(n)
+
+	// Convert our database db.RevokedCertificateInfo types into the pkix representation ready for the
+	// CAS to sign it
+	var revokedCertificates []pkix.RevokedCertificate
+
+	for _, revokedCert := range *revokedList {
+		var sn big.Int
+		sn.SetString(revokedCert.Serial, 10)
+		revokedCertificates = append(revokedCertificates, pkix.RevokedCertificate{
+			SerialNumber:   &sn,
+			RevocationTime: revokedCert.RevokedAt,
+			Extensions:     nil,
+		})
+	}
+
+	// Create a RevocationList representation ready for the CAS to sign
+	// TODO: use a config value for the NextUpdate time duration
+	// TODO: allow SignatureAlgorithm to be specified?
+	revocationList := x509.RevocationList{
+		SignatureAlgorithm:  0,
+		RevokedCertificates: revokedCertificates,
+		Number:              &bn,
+		ThisUpdate:          time.Now().UTC(),
+		NextUpdate:          time.Now().UTC().Add(time.Minute * 10),
```

Comment:
Making the value configuration we might be able to disable (0m) on a "non-master" node if the CA is running distributed using a shared MySQL database.

Reply:
In fact, it might be good to just enable the CRL if configured:
```
// Enabled and generate CRL every 8h
"crl": {"enabled": true, "cacheDuration": "8h"}
// Enabled but do not generate the CRL
"crl": {"enabled": true}
"crl": {"enabled": true, "cacheDuration": "0h"}
// Disabled
// Enabled but do not generate the CRL
"crl": {"enabled": false}
```

Reply:
So this method should be split in two, one like `GetCertificateRevocationList` that returns the CRL from the DB if enabled, and return a 404 or 501 error if not.

And a background job that will start only if enabled and a `cacheDuration` is greater than 0. This job will also be notified with a channel after a revocation.

Reply:
I changed the config slightly to (hopefully) make the behaviour more clear:
`"crl": {"generate": true, "cacheDuration": "8h"}`

`generate` controls whether the automatic generation goroutine is created
`cacheDuration` is also needed on all nodes (if using multiple nodes) so that when a certificate is revoked and the node creates a new CRL, the node has a duration to place in the CRL.

if the `crl` dict is missing, CRL generation/serving is completely disabled.